src/cpuinfo.cpp

```diff
@@ -25,452 +27,721 @@
+    case RISCVExtension::Zksh: {
+        as.SM3P0(t1, t2);
+        break;
+    }
+    case RISCVExtension::Zvbb: {
+        as.VSETIVLI(x0, 1, SEW::E32);
+        as.VCLZ(v1, v2);
+        break;
+    }
+    case RISCVExtension::Zvbc: {
+        SetEGW(as, SEW::E64, 1);
+        as.VCLMUL(v8, v16, v24);
+        break;
+    }
+    case RISCVExtension::Zvkb: {
+        as.VSETIVLI(x0, 1, SEW::E32);
+        as.VANDN(v1, v2, v3);
+        break;
+    }
+    case RISCVExtension::Zvkg: {
+        SetEGW(as, SEW::E32, 4);
+        as.VGHSH(v8, v16, v24);
+        break;
+    }
+    case RISCVExtension::Zvkned: {
+        SetEGW(as, SEW::E32, 4);
+        as.VAESEM_VV(v8, v16);
+        break;
+    }
+    case RISCVExtension::Zvknha: {
+        SetEGW(as, SEW::E32, 4);
+        as.VSHA2MS(v8, v16, v24);
+        break;
+    }
+    case RISCVExtension::Zvknhb: {
+        SetEGW(as, SEW::E64, 4);
+        as.VSHA2MS(v8, v16, v24);
+        break;
+    }
+    case RISCVExtension::Zvksed: {
+        SetEGW(as, SEW::E32, 8);
+        as.VSM4R_VV(v8, v16);
+        break;
+    }
+    case RISCVExtension::Zvksh: {
+        SetEGW(as, SEW::E32, 8);
+        as.VSM3ME(v8, v16, v24);
+        break;
+    }
+    case RISCVExtension::Zfh: {
+        as.FMV_H_X(ft0, x0);
+        as.FMV_H_X(ft1, x0);
+        as.FMV_H_X(ft2, x0);
+        as.FADD_H(ft0, ft1, f2);
+        break;
+    }
+    case RISCVExtension::Zfhmin: {
+        as.FMV_H_X(ft0, x0);
+        break;
+    }
+    case RISCVExtension::Zvfh: {
+        as.VSETIVLI(x0, 1, SEW::E16);
+        as.VMV(v2, x0);
+        as.VMV(v3, x0);
+        as.VFADD(v1, v2, v3);
+        break;
+    }
+    case RISCVExtension::Zvfhmin: {
+        as.VSETIVLI(x0, 1, SEW::E16);
+        as.VMV(v8, x0);
+        as.VFWCVT_F_F(v0, v8);
+        break;
+    }
+    case RISCVExtension::Zfa: {
+        as.FLI_S(ft0, 2.0);
+        break;
+    }
+    case RISCVExtension::Zacas: {
+        as.AMOCAS_D(Ordering::AQRL, t1, t2, t0);
+        break;
+    }
+    case RISCVExtension::Zicond: {
+        as.CZERO_EQZ(t1, t2, t3);
+        break;
+    }
+    case RISCVExtension::Zicntr: {
+        as.RDTIME(t1);
+        break;
+    }
+    case RISCVExtension::Zihpm: {
+        as.CSRR(t1, CSR::HPMCounter3);
+        break;
+    }
+    case RISCVExtension::Zfbfmin: {
+        as.FMV_W_X(ft0, x0);
+        as.FCVT_BF16_S(ft1, ft0);
+        break;
+    }
+    case RISCVExtension::Zicbom: {
+        as.CBO_FLUSH(t0);
+        break;
+    }
+    default: {
+        BISCUIT_ASSERT(false);
+    }
+    }
+}
 
-bool CPUInfo::Has(RISCVExtension extension) const {
-#if defined(__linux__) && defined(__riscv)
-    static const auto [ima, features0] = []() {
+bool CheckExtensionSigill(biscuit::RISCVExtension extension) {
+    using namespace biscuit;
+
+    struct sigaction sa, old_sa;
+    sa.sa_sigaction = signal_handler;
+    sa.sa_flags = SA_SIGINFO;
+    sigemptyset(&sa.sa_mask);
+
+    BISCUIT_ASSERT(sigaction(SIGILL, &sa, &old_sa) == 0);
```

Comment:
Asserts shouldn't have actionable/functional behaviors within them, they should just be sanitizing the results of those functional behaviors. So the sigaction call should be hoisted out of the assert and the return value should be in a variable that's then checked in the assert.